common/scala/src/main/scala/whisk/core/database/couchdb/CouchDbAttachmentStore.scala

```diff
@@ -0,0 +1,171 @@
+        val id = response.fields("id").convertTo[String]
+        val rev = response.fields("rev").convertTo[String]
+        DocInfo ! (id, rev)
+
+      case Left(StatusCodes.NotFound) =>
+        transid
+          .finished(this, start, s"[ATT_PUT] '$dbName' uploading attachment '$name' of document '$doc'; not found")
+        throw NoDocumentException("Not found on 'readAttachment'.")
+
+      case Left(code) =>
+        transid.failed(
+          this,
+          start,
+          s"[ATT_PUT] '$dbName' failed to upload attachment '$name' of document '$doc'; http status '$code'")
+        throw new Exception("Unexpected http response code: " + code)
+    }
+
+    reportFailure(
+      f,
+      failure =>
+        transid.failed(
+          this,
+          start,
+          s"[ATT_PUT] '$dbName' internal error, name: '$name', doc: '$doc', failure: '${failure.getMessage}'",
+          ErrorLevel))
+  }
+
+  override protected[core] def readAttachment[T](doc: DocInfo, name: String, sink: Sink[ByteString, Future[T]])(
+    implicit transid: TransactionId): Future[(ContentType, T)] = {
+
+    val start = transid.started(
+      this,
+      LoggingMarkers.DATABASE_ATT_GET,
+      s"[ATT_GET] '$dbName' finding attachment '$name' of document '$doc'")
+
+    require(doc != null, "doc undefined")
+    require(doc.rev.rev != null, "doc revision must be specified")
+
+    val f = client.getAttachment[T](doc.id.id, doc.rev.rev, name, sink)
+    val g = f.map {
+      case Right((contentType, result)) =>
+        transid.finished(this, start, s"[ATT_GET] '$dbName' completed: found attachment '$name' of document '$doc'")
+        (contentType, result)
+
+      case Left(StatusCodes.NotFound) =>
+        transid.finished(
+          this,
+          start,
+          s"[ATT_GET] '$dbName', retrieving attachment '$name' of document '$doc'; not found.")
+        throw NoDocumentException("Not found on 'readAttachment'.")
+
+      case Left(code) =>
+        transid.failed(
+          this,
+          start,
+          s"[ATT_GET] '$dbName' failed to get attachment '$name' of document '$doc'; http status: '$code'")
+        throw new Exception("Unexpected http response code: " + code)
+    }
+
+    reportFailure(
+      g,
+      failure =>
+        transid.failed(
+          this,
+          start,
+          s"[ATT_GET] '$dbName' internal error, name: '$name', doc: '$doc', failure: '${failure.getMessage}'",
+          ErrorLevel))
+  }
+
+  override protected[core] def deleteAttachments(doc: DocInfo)(implicit transid: TransactionId): Future[Boolean] =
+    // NOTE: this method is not intended for standalone use for CouchDB.
+    // To delete attachments, it is expected that the entire document is deleted.
+    Future.successful(true)
+
+  override def shutdown(): Unit = {
+    Await.ready(client.shutdown(), 1.minute)
+  }
+
+  private def reportFailure[T, U](f: Future[T], onFailure: Throwable => U): Future[T] = {
```

Comment:
could you move report failure to a singleton and use it in the store above also to avoid clone/own?

Reply:
Moved them to `StoreUtils`